mlflow/store/rest_store.py

```diff
@@ -249,6 +250,21 @@ def restore_run(self, run_id):
         req_body = message_to_json(RestoreRun(run_id=run_id))
         self._call_endpoint(RestoreRun, req_body)
 
+    def get_experiment_by_name(self, experiment_name):
+        try:
+            req_body = message_to_json(GetExperimentByName(experiment_name=experiment_name))
+            response_proto = self._call_endpoint(GetExperimentByName, req_body)
+            return Experiment.from_proto(response_proto.experiment)
+        except MlflowException as e:
+            if e.error_code == databricks_pb2.ErrorCode.Name(
+                    databricks_pb2.RESOURCE_DOES_NOT_EXIST):
+                return None
+            # Fall back to using ListExperiments-based implementation.
```

Comment:
Should we check that the error message was that the endpoint does not exists? I don't think we want to fall back on list_experiments for 500s or 429 etc.

Reply:
Good question, we could explicitly raise & not perform the fallback if the server responds with `RESOURCE_LIMIT_EXCEEDED`, but it's hard to raise on `INTERNAL_ERROR`,
since the OSS server responds with XML 404s for nonexistent endpoints (rather than a well-formed JSON exception with an `ENDPOINT_DOES_NOT_EXIST` error code) and we deserialize these client-side as `INTERNAL_ERROR`s. We could get around this using Python `requests` directly & only fall back on 500-level error codes from the server, but other implementors of the OSS REST API might not follow the pattern of using 400-level responses for nonexistent endpoints. Since we only need the fallback logic in one MLflow client version (i.e. we'll add it in MLflow 1.3 and drop it in MLflow 1.4), maybe the easiest thing is to explicitly re-raise on `RESOURCE_LIMIT_EXCEEDED` but fall back to ListExperiments otherwise

Reply:
Are we doing this only for one version? Shouldn't our apis be kept stable (for client and server)?

As for the 404, isn't 404 the standard code to return in case of no such endpoint? I think it's better to not to support stores that do not follow standards than falling back to listExperiments for everybody.

Reply:
Talked to @mateiz and @aarondav about this - we'd like to have a policy of supporting client-server backwards compatibility for some finite set of versions (i.e. we'll keep the client backwards-compatible with old servers for at least 1 minor version, but we won't guarantee that all MLflow 1.x clients work against all MLflow 1.y clients, for x > y), so I might defer adding more complex fallback logic for now.